### DISCUSSION TITLE
Combination of the keywords "this" and "params"

### DISCUSSION THREAD

**Discussion by TobiasKnauss** [dotnet/csharplang#6078] [Language Ideas] [answered]
I suggest to enable and allow the combination of the keywords `this` and `params`.
If the desired functionality was possible, but a combination of the keywords is not desired or possible, a new keyword might be introduced. The name of the keyword may be `this_params`, although it's quite ugly and there are nicer alternatives for sure.

The reason for this suggestion is that I have created some extension methods, e.g. for `Enum` with `[Flags]` attribute:

```
using System;
using System.Linq;

public static class FlagsEnumHelper
{
  public static T MergeFlags<T> (params T[] i_flags)
    where T : struct, Enum
  {
    var type = typeof (T);
    if (!type.IsDefined (typeof (FlagsAttribute), false))
      throw new ArgumentException ($"The given enum type '{type}' does not have the {nameof (FlagsAttribute)}.");

    var zero = (T)Convert.ChangeType (0, Enum.GetUnderlyingType (type));
    var result = i_flags.Aggregate (zero, (a, b) => (dynamic)a | (dynamic)b);
    return result;
  }
}

public static class FlagsEnumHelper2
{
  public static T MergeFlags<T> (this T[] i_flags)
    where T : struct, Enum
  {
    return FlagsEnumHelper.MergeFlags (i_flags);
  }
}
```

You can see that these two methods are in different classes, which is required because the compiler is not able to choose any of these two equivalent methods if they were in the same class.

If the keywords could be combined or a new keyword was given, it would be possible to avoid the second class and have only one.
Example:
```
public static class FlagsEnumHelper
{
  public static T MergeFlags<T> (this params T[] i_flags)
  ...
}
```

**Comment by HaloFour** [comment]
What exactly does `this params` enable?  What does the callsite look like?

**Reply by TobiasKnauss** [reply]
It would be possible to call the method as extension method on an array, and as static method with single elements:

usage as `params[]`:
```
var combinedValue = FlagsEnumHelper.MergeFlags (ExampleEnum.FlagA, ExampleEnum.FlagB, ExampleEnum.FlagC);
```
usage as `this`:
```
var flags = new[] { ExampleEnum.FlagD, ExampleEnum.FlagE, ExampleEnum.FlagF };
var combinedValue = flags.MergeFlags ();
```

**Reply by HaloFour** [reply]
You can accomplish that today by having an extension method on the array of the type.  Yes, that's an extra method to write, but this sounds like a very infrequent thing to want or need to do, so I don't see how it justifies requiring special syntax or handling.  I don't see why you'd want to call `MergeFlags` on a single value of an enum in the first place, so I'd probably just omit that overload entirely.

**Reply by TobiasKnauss** [reply]
If I understand you correctly, you are suggesting the solution that I already use and that I have added as example in the first post, aren't you?

I don't want to call `MergeFlags` or other similar methods on a single value, this of course doesn't make sense. But I would like to have the flexibility to call it using multiple single elements, or calling it on an array. This can only be achieved by creating two separate methods. If there is an easy way and the benefit justifies the required work, it's would remove the need for such boilerplate classes and methods and produce cleaner code.
My idea is that the compiler could allow the combination, because it just creates calls to the same static method in both cases. So why create two separate methods?

**Reply by HaloFour** [reply]
> But I would like to have the flexibility to call it using multiple single elements

What does this look like?

> So why create two separate methods?

Because this is a rare thing to want to do so it doesn't justify a language feature to support it.  It would likely require substantially more code to implement than would ever be saved.

**Comment by TahirAhmadov** [comment, marked as answer]
Other than the cost/benefit ratio not being in favor of this, it's unlikely any investment will happen in existing extension methods, because the news roles/extensions feature is much more important.

Also, why not just use the `new[] { x, y, z }.MergeFlags()` syntax instead of the `params` syntax? It's still an array allocation behind the scenes, so the performance is the same; in some ways, it's even easier to read - no need for the helper class prefix.
